src/Protocols/Axi4/Stream/Axi4Stream.hs

```diff
@@ -0,0 +1,233 @@
+{-|
+Types and instance declarations for the AXI4-stream protocol.
+-}
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-} -- Hashable (Unsigned n)
+
+module Protocols.Axi4.Stream.Axi4Stream where
+
+-- base
+import           Control.DeepSeq (NFData)
+import           Data.Hashable (Hashable, hashWithSalt)
+import qualified Data.Maybe as Maybe
+import           Data.Proxy
+import qualified Prelude as P
+
+-- clash-prelude
+import           Clash.Prelude hiding (take, concat, length)
+import qualified Clash.Prelude as C
+
+-- me
+import           Protocols.Internal
+import qualified Protocols.Df as Df
+import qualified Protocols.DfConv as DfConv
+import           Protocols.Hedgehog.Internal
+
+
+instance (KnownNat n) => Hashable (Unsigned n)
+instance (KnownNat n, Hashable a) => Hashable (Vec n a) where
+  hashWithSalt s v = hashWithSalt s (toList v)
+
+-- | Configuration for AXI4 Stream protocol. Determines the width of some
+-- fields in 'Axi4StreamM2S'.
+data Axi4StreamConfig
+  = Axi4StreamConfig
+  { _dataWidth :: Nat
+  , _idWidth :: Nat
+  , _destWidth :: Nat
+  }
+
+-- | Grab '_dataWidth' at the type level.
+type family DataWidth (conf :: Axi4StreamConfig) where
+  DataWidth ('Axi4StreamConfig a _ _) = a
+
+-- | Grab '_idWidth' at the type level.
+type family IdWidth (conf :: Axi4StreamConfig) where
+  IdWidth ('Axi4StreamConfig _ a _) = a
+
+-- | Grab '_destWidth' at the type level.
+type family DestWidth (conf :: Axi4StreamConfig) where
+  DestWidth ('Axi4StreamConfig _ _ a) = a
+
+-- | Shorthand for a "well-behaved" config, so that we don't need to write out
+-- a bunch of type constraints later. Holds for every configuration; don't worry
+-- about implementing this class.
+type GoodAxi4StreamConfig conf =
+  ( KnownNat (DataWidth conf)
+  , KnownNat (IdWidth conf)
+  , KnownNat (DestWidth conf)
+  )
+
+-- | A byte sent along an AXI4 Stream. Each byte can either be a data byte, a
+-- position byte, or a null byte. The value of position and null bytes should
+-- be ignored. Additionally, null bytes can be added or dropped.
+data Axi4StreamByte
+  = DataByte (Unsigned 8)
+  | PositionByte
+  | NullByte
+  deriving (Generic, C.NFDataX, C.ShowX, Eq, NFData, Show, Hashable)
+
+-- | Data sent from manager to subordinate. The tvalid field is left out: messages with @tvalid = False@
+-- should be sent as a @NoAxi4StreamM2S@. The tdata, @tstrb@, and @tkeep@ fields are
+-- all grouped in the @_tdata@ field in this datatype (see @Axi4StreamByte@).
+data Axi4StreamM2S (conf :: Axi4StreamConfig) (userType :: Type)
+  = NoAxi4StreamM2S
+  | Axi4StreamM2S
+    { _tdata :: Vec (DataWidth conf) Axi4StreamByte
```

Comment:
This is definitely the Clashy way to define the spec, but I'm wondering if we really want this. There are two considerations:

 * How this is represented in HDL? I'm assuming having this at a top entity does not give you an API compatible AXI4 port mapping.
 * If we don't care about the port mapping: wouldn't it be Classier to simply make that `a`?

It seems like the current implementation gives us neither the flexibility we'd really want, nor API compatible HDL. I think I'd prefer to define `_tdata :: a`, with a separate fields for `_tkeep`/`_tstrb`. To create an API compatible component we'd simply need to make sure that `a ~ BitVector n` and the `_tkeep`/`_strb` are set accordingly.

Alternatively, we could define `_tdata :: BitVector n` and assume that most/all Clash designs don't want to deal with AXI4 and would therefore convert them to `Df` at their borders.

What do you think?

Reply:
Good point, I'll add `_tkeep` and `_tstrb` separately